src/ui/Search.js

```diff
@@ -0,0 +1,148 @@
+import React, {Component} from 'react';
+import Modal from 'react-modal';
+import {UP, DOWN, ESC, ENTER, PGUP, PGDN, F3} from '../keycode';
+
+// TODO: Make sure to bind modal to your appElement (http://reactcommunity.org/react-modal/accessibility/)
+// Modal.setAppElement('#yourAppElement');
+
+export default class extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {showSearchModal: false, searchString: '', searchMode: 0};
+    this.searchBoxNode = null;
+  }
+
+  handleKeyModal = e => {
+    if (e.keyCode === ENTER || e.keyCode === ESC) { // enter or escape
+      this.setState({showSearchModal: false});
+
+      // NOTE(Oak): getActiveNode uses document.activeElement, but right now
+      // our focus is on the modal, so we use setTimeout to wait for the modal to close
+      // first
+      setTimeout(() => {
+        if (this.state.searchString === '') return;
+        this.props.searchModes[this.state.searchMode].initSearch(this.state.searchString);
+      });
+
+    } else if (e.keyCode === UP) {
+      if (this.state.searchMode > 0) {
+        this.setState((prevState => ({searchMode: prevState.searchMode - 1})));
+      }
+    } else if (e.keyCode === DOWN) {
+      if (this.state.searchMode < this.props.searchModes.length - 1) {
+        this.setState((prevState => ({searchMode: prevState.searchMode + 1})));
+      }
+    }
+  }
+
+
+  handleKeyGlobal = e => {
+    switch (e.keyCode) {
+    case F3:
+      e.preventDefault(); // prevent the browser search
+
+      this.setState({showSearchModal: true});
+      return;
+
+    case PGUP:
+      e.preventDefault(); // we never want pgup and pgdn to actually do pgup and pgdn
+
+      // NOTE: Oak doesn't think the next line is right. We ought to support
+      // searching in non block mode too, but this's the current behavior
```

Comment:
no - we only support searching in blocks mode

Reply:
Got it.